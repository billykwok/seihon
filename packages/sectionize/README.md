# @seihon/sectionize

[![npm version](https://badgen.net/npm/v/@seihon/sectionize)](https://www.npmjs.com/package/@seihon/sectionize)
[![download](https://badgen.net/npm/dm/@seihon/sectionize)](https://www.npmjs.com/package/@seihon/sectionize)
[![minified size](https://badgen.net/bundlephobia/min/@seihon/sectionize)](https://bundlephobia.com/result?p=@seihon/sectionize)
[![GZip size](https://badgen.net/bundlephobia/minzip/@seihon/sectionize)](https://bundlephobia.com/result?p=@seihon/sectionize)
[![lerna](https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg)](https://lerna.js.org)
[![Greenkeeper badge](https://badges.greenkeeper.io/billykwok/seihon.svg)](https://greenkeeper.io)
[![codecov](https://codecov.io/gh/billykwok/seihon/branch/master/graph/badge.svg)](https://codecov.io/gh/billykwok/seihon)
[![CircleCI](https://circleci.com/gh/billykwok/seihon/tree/master.svg?style=svg)](https://circleci.com/gh/billykwok/seihon/tree/master)

> Seihon【製本】(Bookbinding in Japanese) is a JavaScript toolkit that improves your MDX transformation pipeline. It allows you to quickly transform MDX documents into a collection (like turning codices into book).
>
> More importantly, this toolkit is the last piece of the puzzle that enables automatic generation of CMS-less MDX-based static site.

`@seihon/sectionize` is a [`unified`](https://github.com/unifiedjs/unified) plugin that divides a continuous piece of text into chunks wrapped by a customizable tag.

- To use it with [`@mdx-js/loader`](https://github.com/mdx-js/mdx/tree/master/packages/loader), you can add it to the `remarkPlugins` option.

- To use it with [`unified`](https://github.com/unifiedjs/unified), you just need to place this plugin into the `.use()` pipeline.

> In most occasions, you need to use this plugin together with [`unified`](https://github.com/unifiedjs/unified), or [`@mdx-js/loader`](https://github.com/mdx-js/mdx/tree/master/packages/loader) and [`babel-loader`](https://github.com/babel/babel-loader).

## Usage

This is an example of using `@seihon/section` alone. For a complete example of the entire Seihon toolkit, please refer to the main [`README.md`](https://github.com/billykwok/seihon/blob/master/README.md).

```javascript
// webpack.config.js
import sectionize from '@seihon/sectionize';
// ...
module: {
  rules: [
    {
      test: /\.mdx?$/,
      use: ['babel-loader', '@mdx-js/loader', '@seihon/loader']
    }
    // ...
  ];
}
```

```md
// src/content/posts/introducing-seihon.mdx

# Introducing Seihon

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

## Overview

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.

- Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
- Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

```javascript
// src/components/post.jsx
import React from 'react';
import { MDXProvider } from '@mdx-js/react';
import Article from '../content/posts/introducing-seihon.mdx';

export default function Post() {
  return (
    <MDXProvider>
      <Article />
    </MDXProvider>
  );
}
```

Using the above config, you would obtain the following [`react`](https://reactjs.org/) rendering output.

```html
<section>
  <h1>Introducing Seihon</h1>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
    tempor incididunt ut labore et dolore magna aliqua.
  </p>
</section>
<section>
  <h1>Overview</h1>
  <p>
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
    aliquip ex ea commodo consequat.
  </p>
</section>
<section>
  <ul>
    <li>
      Duis aute irure dolor in reprehenderit in voluptate velit esse cillum
      dolore eu fugiat nulla pariatur.
    </li>
    <li>
      Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia
      deserunt mollit anim id est laborum.
    </li>
  </ul>
</section>
```

## Support

This library has been continuously used in many of my personal projects, and is regarded as production-ready. In the foreseeable future, I will continuously maintain and support this library.

## Issues and Feedback

Please voice your opinion and report bugs in the [issues](https://github.com/billykwok/seihon/issues) sections of this GitHub project.

## Contributing

You are more than welcome to add more functionalities, improve documentation, fix bugs, and anything you think is needed. The build step is pretty self-explanatory. Please refer to [CONTRIBUTING.md](https://github.com/billykwok/seihon/blob/master/CONTRIBUTING.md) for more details.

## License

MIT
