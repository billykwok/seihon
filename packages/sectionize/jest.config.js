module.exports = {
  displayName: '@seihon/sectionize',
  testEnvironment: 'node',
  snapshotSerializers: ['jest-serializer-html'],
  coverageProvider: 'v8',
  verbose: true,
};
