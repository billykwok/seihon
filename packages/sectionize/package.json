{
  "name": "@seihon/sectionize",
  "description": "All-in-one responsive image processing library with webpage loader, babel macro and React component",
  "version": "0.0.9",
  "main": "lib/index.js",
  "author": "Billy Kwok <xobkwok@gmail.com>",
  "license": "MIT",
  "keywords": [
    "seihon",
    "mdx",
    "webpack",
    "remark"
  ],
  "files": [
    "lib",
    "types"
  ],
  "types": "types/index.d.ts",
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/billykwok/seihon/packages/sectionize",
  "bugs": {
    "url": "https://github.com/billykwok/seihon/issues"
  },
  "homepage": "https://github.com/billykwok/seihon",
  "scripts": {
    "clean": "rm -rf lib",
    "build": "yarn clean && babel src --root-mode=upward --out-dir lib --extensions .ts,.tsx,.mjs,.js,.jsx",
    "eslint": "eslint ./src/**",
    "dtslint": "dtslint --onlyTestTsNext --localTs ../../node_modules/typescript/lib types",
    "lint": "yarn eslint && yarn dtslint",
    "test": "jest"
  },
  "dependencies": {
    "deepmerge": "^4.2.2",
    "unist-util-find-after": "^2.0.4",
    "unist-util-visit-parents": "^3.0.1"
  }
}
