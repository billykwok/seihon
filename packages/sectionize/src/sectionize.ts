import findAfter from 'unist-util-find-after';
import { Node, Parent } from 'unist';

export default function sectionize<T extends Node, S extends Parent>(
  node: T,
  ancestors: S[],
  tagName: string,
  whitelist: string[]
) {
  const start = node;
  const parent = ancestors[ancestors.length - 1];

  const isEnd = (n: T) => whitelist.indexOf(n.type) < 0;
  const end = findAfter(parent, start, isEnd);

  const startIndex = parent.children.indexOf(start);
  const endIndex = parent.children.indexOf(end);

  const between = parent.children.slice(
    startIndex,
    endIndex > 0 ? endIndex : undefined
  );

  const section = {
    type: 'section',
    children: between,
    data: { hName: tagName }
  };

  parent.children.splice(startIndex, section.children.length, section);
}
