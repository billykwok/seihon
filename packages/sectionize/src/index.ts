import visit from 'unist-util-visit-parents';
import deepmerge from 'deepmerge';
import { Node, Parent } from 'unist';

import defaultOptions from './defaultOptions';
import isFirstInSection from './isFirstInSection';
import sectionize from './sectionize';

export default function plugin<T extends Node>(options = defaultOptions) {
  return (tree: T) => {
    const { tagName, whitelist } = deepmerge(defaultOptions, options);
    visit(tree, isFirstInSection(whitelist), (node, ancestors) =>
      sectionize(node, ancestors as Parent[], tagName, whitelist)
    );
  };
}
