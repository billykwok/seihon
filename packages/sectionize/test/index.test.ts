import { describe, test, expect } from '@jest/globals';
import unified from 'unified';
import remark from 'remark-parse';
import remark2rehype from 'remark-rehype';
import html from 'rehype-stringify';
import prettier from 'prettier';
import { source } from 'common-tags';
import sectionize from '../src';

function process(options?: any) {
  return (literals: TemplateStringsArray, ...placeholders: any[]) => {
    return unified()
      .use(remark)
      .use(sectionize, options)
      .use(remark2rehype)
      .use(html)
      .processSync(
        prettier.format(source(literals, ...placeholders), {
          parser: 'markdown',
        })
      )
      .toString();
  };
}

describe('sectionize', () => {
  test('regular layout', () => {
    expect(
      process()`
        # Title

        Overview

        ## Subheading 1

        Lorem Ipsum

        [Img](https://google.com/abc.png)

        ## Subheading 2

        Lorem Ipsum

        ## Subheading 3

        * item 1
        * item 2
      `
    ).toMatchSnapshot();
  });

  test('layout with consecutive paragraphs and images', () => {
    expect(
      process()`
        # Title

        Overview

        ## Subheading 1

        Lorem Ipsum

        [Img1](https://google.com/abc.png)
        [Img2](https://google.com/abc.png)

        ## Subheading 2

        Lorem Ipsum 1

        Lorem Ipsum 2

        ## Subheading 3

        * item 1
        * item 2
      `
    ).toMatchSnapshot();
  });

  test('layout with custom tag', () => {
    expect(
      process({ tagName: 'custom' })`
        # Title

        Overview

        ## Subheading 1

        Lorem Ipsum

        [Img](https://google.com/abc.png)

        ## Subheading 2

        Lorem Ipsum

        ## Subheading 3

        * item 1
        * item 2
      `
    ).toMatchSnapshot();
  });

  test('layout with whitelist elements', () => {
    expect(
      process({ whitelist: ['heading'] })`
        # Title

        Overview

        ## Subheading 1

        Lorem Ipsum

        [Img](https://google.com/abc.png)

        * item 1
        * item 2
      `
    ).toMatchSnapshot();
  });

  test('layout with customed whitelist elements 1', () => {
    expect(
      process({ whitelist: ['heading', 'blockquote'] })`
        Overview

        > Lorem Ipsum

        Lorem Ipsum

        [Img](https://google.com/abc.png)

        * item 1
        * item 2
      `
    ).toMatchSnapshot();
  });

  test('layout with customed whitelist elements 2', () => {
    expect(
      process({ whitelist: ['heading', 'blockquote'] })`
        > Lorem Ipsum

        Lorem Ipsum

        [Img](https://google.com/abc.png)

        * item 1
        * item 2
      `
    ).toMatchSnapshot();
  });
});
