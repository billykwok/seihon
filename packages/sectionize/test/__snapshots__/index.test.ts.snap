// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sectionize layout with consecutive paragraphs and images 1`] = `
<section>
  <h1>
    Title
  </h1>
  <p>
    Overview
  </p>
</section>
<section>
  <h2>
    Subheading 1
  </h2>
  <p>
    Lorem Ipsum
  </p>
  <p>
    <a href="https://google.com/abc.png">
      Img1
    </a>
    <a href="https://google.com/abc.png">
      Img2
    </a>
  </p>
</section>
<section>
  <h2>
    Subheading 2
  </h2>
  <p>
    Lorem Ipsum 1
  </p>
  <p>
    Lorem Ipsum 2
  </p>
</section>
<section>
  <h2>
    Subheading 3
  </h2>
  <ul>
    <li>
      item 1
    </li>
    <li>
      item 2
    </li>
  </ul>
</section>
`;

exports[`sectionize layout with custom tag 1`] = `
<custom>
  <h1>
    Title
  </h1>
  <p>
    Overview
  </p>
</custom>
<custom>
  <h2>
    Subheading 1
  </h2>
  <p>
    Lorem Ipsum
  </p>
  <p>
    <a href="https://google.com/abc.png">
      Img
    </a>
  </p>
</custom>
<custom>
  <h2>
    Subheading 2
  </h2>
  <p>
    Lorem Ipsum
  </p>
</custom>
<custom>
  <h2>
    Subheading 3
  </h2>
  <ul>
    <li>
      item 1
    </li>
    <li>
      item 2
    </li>
  </ul>
</custom>
`;

exports[`sectionize layout with whitelist elements 1`] = `
<section>
  <h1>
    Title
  </h1>
  <p>
    Overview
  </p>
  <h2>
    Subheading 1
  </h2>
  <p>
    Lorem Ipsum
  </p>
  <p>
    <a href="https://google.com/abc.png">
      Img
    </a>
  </p>
  <ul>
    <li>
      item 1
    </li>
    <li>
      item 2
    </li>
  </ul>
</section>
`;

exports[`sectionize regular layout 1`] = `
<section>
  <h1>
    Title
  </h1>
  <p>
    Overview
  </p>
</section>
<section>
  <h2>
    Subheading 1
  </h2>
  <p>
    Lorem Ipsum
  </p>
  <p>
    <a href="https://google.com/abc.png">
      Img
    </a>
  </p>
</section>
<section>
  <h2>
    Subheading 2
  </h2>
  <p>
    Lorem Ipsum
  </p>
</section>
<section>
  <h2>
    Subheading 3
  </h2>
  <ul>
    <li>
      item 1
    </li>
    <li>
      item 2
    </li>
  </ul>
</section>
`;
